### ISSUE TITLE
Failed preserve method context when destructuring an object in for-in cycle in if without brackets

### ISSUE THREAD

**Issue by edoardocavazza** [babel/babel#6559]
Choose one: is this a bug report or feature request?

**Bug**

### Input Code

https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&code_lz=MYGwhgzhAEBi0G8BQ1XQgUwC4FcAOAFBgJSIpoUCWAZtAVgBaUQB0GAbhgHZYAyzWbhgBOEYuQqTo1APbC6wGVwhZoqylzVNWHbljFkpRtIuWqw0ALxbmbTjwgBtLAF0WAWzCECYUpYB8hsbBqKYq0I5cADTQMi5WaiwQeCCUWAQA5ADUGcQA3BIhaAC--YUhAJCMtmAAJrU-ZUXQxYWtqOQVmLh4AExEpMjGNHTVOvZ8AkKig-VGsvIEYarqmmN2egZDzSZK4RbW67oOzm6e3r5Wgds7kssR0bHxh0kpaZk5TbctX81V2iw6g1fAVmu0jODUK1ikA&debug=false&circleciRepo=&evaluate=false&lineWrap=true&presets=es2015&targets=&version=6.26.0

### Babel Configuration (.babelrc, package.json, cli command)

```js
{
        exclude: [],
        compact: false,
        presets: [
            ['env', {
                targets: {
                    browsers: ['ie >= 11', 'safari >= 8'],
                },
                modules: false,
            }],
        ]
}
```

### Expected Behavior

Transpiled code should store the `this` context before running the `loop` function. (`setup2` method)

### Current Behavior

Transpiled code does not store the `this` context before running the `loop` function. (`setup` method)

### Context

Bundling a third party library with `rollup` and `rollup-plugin-babel`.

### Your Environment

| software         | version(s)
| ---------------- | -------
| Babel            |  6.26.0
| node             | 8.7.0
| npm              | 5.4.2
| Operating System | MacOS

**Comment by Andarist** [comment]
Thanks for reporting this, although it is fixed in babel7 already which is currently in beta. We do not plan to do any backports to v6 (except security ones maybe).

https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&code_lz=MYGwhgzhAEBi0G8BQ1XQgUwC4FcAOAFBgJSIpoUCWAZtAVgBaUQB0GAbhgHZYAyzWbhgBOEYuQqTo1APbC6wGVwhZoqylzVNWHbljFkpRtIuWqw0ALxbmbTjwgBtLAF0WAWzCECYUpYB8hsbBqKYq0I5cADTQMi5WaiwQeCCUWAQA5ADUGcQA3BIhaAC--YUhAJCMtmAAJrU-ZUXQxYWtqOQVmLh4AExEpMjGNHTVOvZ8AkKig-VGsvIEYarqmmN2egZDzSZK4RbW67oOzm6e3r5Wgds7kssR0bHxh0kpaZk5TbctX81V2iw6g1fAVmu0jODUK1ikA&debug=false&circleciRepo=&evaluate=false&lineWrap=true&presets=es2015&targets=&version=7.0.0-beta.3

**Comment by Andarist** [comment]
Sorry for closing for a brief moment, kinda still sleepy - going for a ☕️ 

I have looked only at one of the loops which works fine, but the other (as reported) is indeed broken.

**Comment by nveenjain** [comment]
I would like to work on this. @Andarist, I agree, but i don't know how/where/which package to start. Could you please guide me?

**Comment by Andarist** [comment]
@nveenjain sure thing! Im not entirely sure, but my guess is that you need to look into `babel-plugin-transform-block-scoping`. To confirm just take the input code and transform it only with that single plugin to see if it is the one which transforms this.

If you have any questions you can always join us at slack and ask questions there or ping me directly there too :)

**Comment by jukben** [comment]
@Andarist I've looked on this issue but It seems to me that it's fixed in 7.0.0-beta.3, doesn't it?

`var _this2 = this;` is persisted

**Comment by Andarist** [comment]
@jukben you are right, ive probably looked on the babel@6.26.0 output instead of the babel@7.0.0-beta.3. Thanks for checking!

As the issue seems to be fixed in latest beta, Im closing this issue.
